core/src/main/scala/org/apache/spark/BarrierTaskContext.scala

```diff
@@ -283,6 +284,7 @@ object BarrierTaskContext {
   @Since("2.4.0")
   def get(): BarrierTaskContext = TaskContext.get().asInstanceOf[BarrierTaskContext]
 
-  private val timer = new Timer("Barrier task timer for barrier() calls.")
+  private val timer = ThreadUtils.newSingleThreadScheduledExecutor(
```

Comment:
Hm I guess there is no place to shut this down. It's not critical and not something we have to change here. I wonder if it should be a daemon thread though? seems weird to spawn a thread that nothing in the code can terminate

Reply:
The origin Timer is not a daemon thread, so I followed the same behavior.
There are some code used to cancel the timer task. 
https://github.com/apache/spark/blob/db8c3058fb9d8e75f03dea6a821c50462315c513/core/src/main/scala/org/apache/spark/BarrierTaskContext.scala#L120

Reply:
That doesn't shut down the executor though, and the other code paths do. I don't see how the non-daemon thread ever terminates. It may not matter as the Spark context lives as long as the app. It'd be tidy to shut it down, or make it a daemon anyway, but not strictly related to the change you're making

Reply:
Got it. Let's change it to daemon thread.